File: node/src/services/reportService.ts
/**
 * Security and dependency report generation
 */
import { spawnSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import { SecurityReport, DependencyReport, ComprehensiveReport } from "../types";
import { log, writeLog, getLogDir } from "../utils/logger";
import { getOutdatedPackages } from "./packageService";
import { findCircularDependencies } from "./dependencyService";

/**
 * Generate security audit report
 * @param projectPath Path to project
 * @returns Security report data
 */
export const generateSecurityReport = (projectPath: string): SecurityReport => {
  const auditResult = spawnSync("npm", ["audit", "--json"], { cwd: projectPath, encoding: "utf-8" });
  try {
    return JSON.parse(auditResult.stdout || "{}");
  } catch {
    return {};
  }
};

/**
 * Generate dependency tree report
 * @param projectPath Path to project
 * @returns Dependency report data
 */
export const generateDependencyReport = (projectPath: string): DependencyReport => {
  const lsResult = spawnSync("npm", ["ls", "--json", "--all"], { cwd: projectPath, encoding: "utf-8" });
  try {
    return JSON.parse(lsResult.stdout || "{}");
  } catch {
    return {};
  }
};

/**
 * Check for breaking changes in package updates
 * @param packageName Package to check
 * @param currentVersion Current version
 * @param latestVersion Latest version
 * @returns Breaking change analysis
 */
const checkBreakingChanges = (packageName: string, currentVersion: string, latestVersion: string): any => {
  // Check if major version change (likely breaking)
  const currentMajor = parseInt(currentVersion.split('.')[0]);
  const latestMajor = parseInt(latestVersion.split('.')[0]);
  const hasMajorChange = latestMajor > currentMajor;
  
  // Get package info for changelog analysis
  const infoResult = spawnSync("npm", ["info", packageName, "--json"], { encoding: "utf-8" });
  let packageInfo = {};
  try {
    packageInfo = JSON.parse(infoResult.stdout || "{}");
  } catch {}
  
  return {
    hasMajorVersionChange: hasMajorChange,
    riskLevel: hasMajorChange ? 'high' : 'low',
    changelog: (packageInfo as any).homepage || '',
    repository: (packageInfo as any).repository?.url || '',
    hasBreakingChanges: hasMajorChange, // Simplified: major version = breaking changes
    migrationRequired: hasMajorChange
  };
};

/**
 * Check peer dependency compatibility
 * @param packageName Package to check
 * @param projectPath Project path
 * @returns Peer dependency analysis
 */
const checkPeerDependencies = (packageName: string, projectPath: string): any => {
  const infoResult = spawnSync("npm", ["info", packageName, "peerDependencies", "--json"], { 
    cwd: projectPath, 
    encoding: "utf-8" 
  });
  
  try {
    const peerDeps = JSON.parse(infoResult.stdout || "{}");
    return {
      hasPeerDependencies: Object.keys(peerDeps).length > 0,
      peerDependencies: peerDeps,
      compatibilityIssues: [] // Simplified for now
    };
  } catch {
    return {
      hasPeerDependencies: false,
      peerDependencies: {},
      compatibilityIssues: []
    };
  }
};

/**
 * Analyze breaking changes for all outdated packages
 * @param outdatedPackages Outdated packages to analyze
 * @param projectPath Project path
 * @returns Breaking change analysis
 */
const analyzeBreakingChanges = (outdatedPackages: Record<string, any>, projectPath: string): any => {
  const analysis = {
    safeUpdates: [] as string[],
    riskyUpdates: [] as string[],
    breakingChanges: {} as Record<string, any>,
    peerDependencyIssues: {} as Record<string, any>
  };
  
  for (const [packageName, details] of Object.entries(outdatedPackages)) {
    const breakingAnalysis = checkBreakingChanges(packageName, details.current, details.latest);
    const peerAnalysis = checkPeerDependencies(packageName, projectPath);
    
    analysis.breakingChanges[packageName] = breakingAnalysis;
    analysis.peerDependencyIssues[packageName] = peerAnalysis;
    
    // Categorize as safe or risky
    if (breakingAnalysis.riskLevel === 'low' && !peerAnalysis.hasPeerDependencies) {
      analysis.safeUpdates.push(packageName);
    } else {
      analysis.riskyUpdates.push(packageName);
    }
  }
  
  return analysis;
};

/**
 * Get outdated packages for reporting (read-only, no updates)
 * @param projectPath Path to project
 * @returns Record of outdated packages
 */
const getOutdatedPackagesReadOnly = (projectPath: string): Record<string, any> => {
  const result = spawnSync("npm", ["outdated", "--json"], { cwd: projectPath, encoding: "utf-8" });
  
  if (result.stdout) {
    try {
      return JSON.parse(result.stdout);
    } catch {
      return {};
    }
  }
  return {};
};

/**
 * Generate comprehensive security and dependency report
 * @param projectPath Path to project
 */
export const generateComprehensiveReport = (projectPath: string): void => {
  log("\n=== Generating Comprehensive Security & Dependency Report ===");
  
  // Gather all report data (READ-ONLY)
  const securityReport = generateSecurityReport(projectPath);
  const dependencyReport = generateDependencyReport(projectPath);
  const outdatedPackages = getOutdatedPackagesReadOnly(projectPath);
  const breakingChangeAnalysis = analyzeBreakingChanges(outdatedPackages, projectPath);
  
  // Build comprehensive report
  const report: ComprehensiveReport = {
    timestamp: new Date().toISOString(),
    project: projectPath,
    security: {
      vulnerabilities: securityReport.vulnerabilities || {},
      summary: securityReport.metadata || {},
      vulnerable_packages: Object.keys(securityReport.vulnerabilities || {})
    },
    dependencies: {
      total: Object.keys(dependencyReport.dependencies || {}).length,
      circular: findCircularDependencies(dependencyReport),
      outdated: Object.keys(outdatedPackages).length,
      outdated_list: outdatedPackages
    },
    breakingChanges: {
      safeUpdates: breakingChangeAnalysis.safeUpdates,
      riskyUpdates: breakingChangeAnalysis.riskyUpdates,
      analysis: breakingChangeAnalysis.breakingChanges,
      peerDependencyIssues: breakingChangeAnalysis.peerDependencyIssues
    },
    recommendations: []
  };
  
  // Generate recommendations
  generateRecommendations(report);
  
  // Save report to file
  const reportFile = saveReportToFile(report);
  
  // Display summary
  displayReportSummary(report, reportFile);
  
  writeLog(`Comprehensive report generated: ${reportFile}`);
};

/**
 * Generate actionable recommendations based on report data
 */
const generateRecommendations = (report: ComprehensiveReport): void => {
  if (report.security.vulnerable_packages.length > 0) {
    report.recommendations.push("Run with --security-only to update vulnerable packages");
  }
  if (report.dependencies.circular.length > 0) {
    report.recommendations.push("Review circular dependencies for potential refactoring");
  }
  if (report.dependencies.outdated > 0) {
    report.recommendations.push("Consider updating outdated packages with --minor-only for safer updates");
  }
  if (report.breakingChanges.safeUpdates.length > 0) {
    report.recommendations.push(`${report.breakingChanges.safeUpdates.length} packages can be safely updated without breaking changes`);
  }
  if (report.breakingChanges.riskyUpdates.length > 0) {
    report.recommendations.push(`${report.breakingChanges.riskyUpdates.length} packages may have breaking changes - review before updating`);
  }
};

/**
 * Save report to JSON file
 */
const saveReportToFile = (report: ComprehensiveReport): string => {
  const reportFile = path.join(getLogDir(), `security-report-${new Date().toISOString().replace(/[:.]/g, '-')}.json`);
  fs.writeFileSync(reportFile, JSON.stringify(report, null, 2));
  return reportFile;
};

/**
 * Display formatted report summary to console
 */
const displayReportSummary = (report: ComprehensiveReport, reportFile: string): void => {
  log(`\n📊 SECURITY & DEPENDENCY REPORT`);
  log(`📁 Project: ${report.project}`);
  log(`🔒 Vulnerabilities: ${report.security.vulnerable_packages.length}`);
  log(`📦 Total Dependencies: ${report.dependencies.total}`);
  log(`🔄 Circular Dependencies: ${report.dependencies.circular.length}`);
  log(`⚠️  Outdated Packages: ${report.dependencies.outdated}`);
  
  // Breaking changes summary
  log(`\n🔍 BREAKING CHANGE ANALYSIS`);
  log(`✅ Safe Updates: ${report.breakingChanges.safeUpdates.length}`);
  log(`⚠️  Risky Updates: ${report.breakingChanges.riskyUpdates.length}`);
  
  if (report.security.vulnerable_packages.length > 0) {
    log(`\n🚨 VULNERABLE PACKAGES:`);
    report.security.vulnerable_packages.forEach((pkg: string) => log(`  - ${pkg}`));
  }
  
  if (report.dependencies.circular.length > 0) {
    log(`\n🔄 CIRCULAR DEPENDENCIES:`);
    report.dependencies.circular.forEach((cycle: string) => log(`  - ${cycle}`));
  }
  
  if (report.breakingChanges.safeUpdates.length > 0) {
    log(`\n✅ SAFE UPDATES (No Breaking Changes):`);
    report.breakingChanges.safeUpdates.forEach((pkg: string) => log(`  - ${pkg}`));
  }
  
  if (report.breakingChanges.riskyUpdates.length > 0) {
    log(`\n⚠️  RISKY UPDATES (Potential Breaking Changes):`);
    report.breakingChanges.riskyUpdates.forEach((pkg: string) => log(`  - ${pkg}`));
  }
  
  log(`\n💡 RECOMMENDATIONS:`);
  report.recommendations.forEach((rec: string) => log(`  - ${rec}`));
  
  log(`\n📄 Full report saved: ${reportFile}`);
};

/**
 * Get safe packages for priority updating
 * @param outdatedPackages Already fetched outdated packages
 * @returns Array of safe package names
 */
export const getSafePackagesForUpdate = (outdatedPackages: Record<string, any>, projectPath: string): string[] => {
  const breakingChangeAnalysis = analyzeBreakingChanges(outdatedPackages, projectPath);
  return breakingChangeAnalysis.safeUpdates;
};

File: python/packUpdate/services/report_service.py
"""
Security and dependency report generation
"""
import subprocess
import json
import os
from datetime import datetime
from ..utils.logger import log, write_log, get_log_dir
from .package_service import get_outdated_packages

def generate_security_report(project_path):
    """Generate security audit report."""
    try:
        result = subprocess.run(['npm', 'audit', '--json'], cwd=project_path, capture_output=True, text=True)
        return json.loads(result.stdout) if result.stdout.strip() else {}
    except:
        return {}

def generate_dependency_report(project_path):
    """Generate dependency tree report."""
    try:
        result = subprocess.run(['npm', 'ls', '--json', '--all'], cwd=project_path, capture_output=True, text=True)
        return json.loads(result.stdout) if result.stdout.strip() else {}
    except:
        return {}

def find_circular_dependencies(deps, visited=None, path=None):
    """Find circular dependencies in the dependency tree."""
    if visited is None:
        visited = set()
    if path is None:
        path = []
    
    circular = []
    if not deps or 'dependencies' not in deps:
        return circular
    
    for name, info in deps['dependencies'].items():
        if name in visited:
            if name in path:
                circular.append(' → '.join(path + [name]))
            continue
        visited.add(name)
        circular.extend(find_circular_dependencies(info, visited, path + [name]))
    
    return circular

def check_breaking_changes(package_name, current_version, latest_version):
    """Check for breaking changes in package updates"""
    # Check if major version change (likely breaking)
    try:
        current_major = int(current_version.split('.')[0])
        latest_major = int(latest_version.split('.')[0])
        has_major_change = latest_major > current_major
    except:
        has_major_change = False
    
    # Get package info for changelog analysis
    try:
        result = subprocess.run(['npm', 'info', package_name, '--json'], capture_output=True, text=True)
        package_info = json.loads(result.stdout) if result.stdout.strip() else {}
    except:
        package_info = {}
    
    return {
        'hasMajorVersionChange': has_major_change,
        'riskLevel': 'high' if has_major_change else 'low',
        'changelog': package_info.get('homepage', ''),
        'repository': package_info.get('repository', {}).get('url', '') if isinstance(package_info.get('repository'), dict) else '',
        'hasBreakingChanges': has_major_change,
        'migrationRequired': has_major_change
    }

def check_peer_dependencies(package_name, project_path):
    """Check peer dependency compatibility"""
    try:
        result = subprocess.run(['npm', 'info', package_name, 'peerDependencies', '--json'], 
                              cwd=project_path, capture_output=True, text=True)
        peer_deps = json.loads(result.stdout) if result.stdout.strip() else {}
    except:
        peer_deps = {}
    
    return {
        'hasPeerDependencies': len(peer_deps) > 0,
        'peerDependencies': peer_deps,
        'compatibilityIssues': []  # Simplified for now
    }

def analyze_breaking_changes(outdated_packages, project_path):
    """Analyze breaking changes for all outdated packages"""
    analysis = {
        'safeUpdates': [],
        'riskyUpdates': [],
        'breakingChanges': {},
        'peerDependencyIssues': {}
    }
    
    for package_name, details in outdated_packages.items():
        breaking_analysis = check_breaking_changes(package_name, details.get('current', ''), details.get('latest', ''))
        peer_analysis = check_peer_dependencies(package_name, project_path)
        
        analysis['breakingChanges'][package_name] = breaking_analysis
        analysis['peerDependencyIssues'][package_name] = peer_analysis
        
        # Categorize as safe or risky
        if breaking_analysis['riskLevel'] == 'low' and not peer_analysis['hasPeerDependencies']:
            analysis['safeUpdates'].append(package_name)
        else:
            analysis['riskyUpdates'].append(package_name)
    
    return analysis

def generate_comprehensive_report(project_path):
    """Generate comprehensive security and dependency report."""
    log("\n=== Generating Comprehensive Security & Dependency Report ===")
    
    security_report = generate_security_report(project_path)
    dependency_report = generate_dependency_report(project_path)
    outdated_packages = get_outdated_packages(project_path)
    breaking_change_analysis = analyze_breaking_changes(outdated_packages, project_path)
    
    report = {
        'timestamp': datetime.now().isoformat(),
        'project': project_path,
        'security': {
            'vulnerabilities': security_report.get('vulnerabilities', {}),
            'summary': security_report.get('metadata', {}),
            'vulnerable_packages': list(security_report.get('vulnerabilities', {}).keys())
        },
        'dependencies': {
            'total': len(dependency_report.get('dependencies', {})),
            'circular': find_circular_dependencies(dependency_report),
            'outdated': len(outdated_packages),
            'outdated_list': outdated_packages
        },
        'breakingChanges': {
            'safeUpdates': breaking_change_analysis['safeUpdates'],
            'riskyUpdates': breaking_change_analysis['riskyUpdates'],
            'analysis': breaking_change_analysis['breakingChanges'],
            'peerDependencyIssues': breaking_change_analysis['peerDependencyIssues']
        },
        'recommendations': []
    }
    
    # Add recommendations
    if report['security']['vulnerable_packages']:
        report['recommendations'].append("Run with --security-only to update vulnerable packages")
    if report['dependencies']['circular']:
        report['recommendations'].append("Review circular dependencies for potential refactoring")
    if report['dependencies']['outdated'] > 0:
        report['recommendations'].append("Consider updating outdated packages with --minor-only for safer updates")
    if report['breakingChanges']['safeUpdates']:
        report['recommendations'].append(f"{len(report['breakingChanges']['safeUpdates'])} packages can be safely updated without breaking changes")
    if report['breakingChanges']['riskyUpdates']:
        report['recommendations'].append(f"{len(report['breakingChanges']['riskyUpdates'])} packages may have breaking changes - review before updating")
    
    report_file = os.path.join(get_log_dir(), f"security-report-{datetime.now().isoformat().replace(':', '-').replace('.', '-')}.json")
    if not os.path.exists(get_log_dir()):
        os.makedirs(get_log_dir(), exist_ok=True)
    
    with open(report_file, 'w') as f:
        json.dump(report, f, indent=2)
    
    # Display summary
    display_report_summary(report, report_file)
    write_log(f"Comprehensive report generated: {report_file}")

def display_report_summary(report, report_file):
    """Display formatted report summary to console"""
    log(f"\n📊 SECURITY & DEPENDENCY REPORT")
    log(f"📁 Project: {report['project']}")
    log(f"🔒 Vulnerabilities: {len(report['security']['vulnerable_packages'])}")
    log(f"📦 Total Dependencies: {report['dependencies']['total']}")
    log(f"🔄 Circular Dependencies: {len(report['dependencies']['circular'])}")
    log(f"⚠️  Outdated Packages: {report['dependencies']['outdated']}")
    
    # Breaking changes summary
    log(f"\n🔍 BREAKING CHANGE ANALYSIS")
    log(f"✅ Safe Updates: {len(report['breakingChanges']['safeUpdates'])}")
    log(f"⚠️  Risky Updates: {len(report['breakingChanges']['riskyUpdates'])}")
    
    if report['security']['vulnerable_packages']:
        log(f"\n🚨 VULNERABLE PACKAGES:")
        for pkg in report['security']['vulnerable_packages']:
            log(f"  - {pkg}")
    
    if report['dependencies']['circular']:
        log(f"\n🔄 CIRCULAR DEPENDENCIES:")
        for cycle in report['dependencies']['circular']:
            log(f"  - {cycle}")
    
    if report['breakingChanges']['safeUpdates']:
        log(f"\n✅ SAFE UPDATES (No Breaking Changes):")
        for pkg in report['breakingChanges']['safeUpdates']:
            log(f"  - {pkg}")
    
    if report['breakingChanges']['riskyUpdates']:
        log(f"\n⚠️  RISKY UPDATES (Potential Breaking Changes):")
        for pkg in report['breakingChanges']['riskyUpdates']:
            log(f"  - {pkg}")
    
    log(f"\n💡 RECOMMENDATIONS:")
    for rec in report['recommendations']:
        log(f"  - {rec}")
    
    log(f"\n📄 Full report saved: {report_file}")

def get_safe_packages_for_update(project_path):
    """Get safe packages for priority updating"""
    outdated_packages = get_outdated_packages(project_path)
    breaking_change_analysis = analyze_breaking_changes(outdated_packages, project_path)
    return breaking_change_analysis['safeUpdates']
